ethcore/src/blockchain/blockchain.rs

```diff
@@ -452,40 +355,26 @@ impl BlockChain {
 	///   ```json
 	///   { blocks: [B4, B3, A3, A4], ancestor: A2, index: 2 }
 	///   ```
-	pub fn tree_route(&self, from: H256, to: H256) -> Option<TreeRoute> {
-		let from_details = match self.block_details(&from) {
-			Some(h) => h,
-			None => return None,
-		};
-		let to_details = match self.block_details(&to) {
-			Some(h) => h,
-			None => return None,
-		};
-		Some(self.tree_route_aux((&from_details, &from), (&to_details, &to)))
-	}
-
-	/// Similar to `tree_route` function, but can be used to return a route
-	/// between blocks which may not be in database yet.
-	fn tree_route_aux(&self, from: (&BlockDetails, &H256), to: (&BlockDetails, &H256)) -> TreeRoute {
+	pub fn tree_route(&self, from: H256, to: H256) -> TreeRoute {
 		let mut from_branch = vec![];
 		let mut to_branch = vec![];
 
-		let mut from_details = from.0.clone();
-		let mut to_details = to.0.clone();
-		let mut current_from = from.1.clone();
-		let mut current_to = to.1.clone();
+		let mut from_details = self.block_details(&from).expect(&format!("0. Expected to find details for block {:?}", from));
```

Comment:
asserting on an external factor such as what is or is not in some particular portion of the database is not particularly defensive coding.

the only proper use of panics/assertions in live code is for specifying logical certainties (`assert!(1 != 0);`) or absolute assertions over the high-level design that are themselves well tested (e.g. triemutdb asserts the root is in the database).

it is sometimes reasonable to use them to specify postconditions when used in order to specify a function's preconditions, this in this case they should be simple enough that it can be done up front in the code prior to the implementation proper.

if you cannot point to comprehensive tests and the structure of the software's design that means your assertion _can never be_ false, then it should not panic, but rather fail-over at a higher level.

in this case, there is a mere "expectation" that the database happens to have a particular entry in it, derived from a potentially arbitrary other number of entries, themselves each populated by code in other modules, and all taken from an argument of unknown provenance.

in short, unless you are willing to bet the farm on this function _never_ being used "improperly", and can provide that by design it cannot happen, then a panic is unacceptable in this position.